### PR TITLE
Initial split of diff computation from breaking change computation

### PR DESCRIPTION
**Pull request by melinath**
fixes b/295223615

I tried to keep this as small as possible but it ended up much larger than I expected. However, most of the length is test cases.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: tools/diff-processor/diff/diff.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+	Old *schema.Resource
+	New *schema.Resource
+}
+
+type FieldDiff struct {
+	Old *schema.Schema
+	New *schema.Schema
+}
+
+func ComputeSchemaDiff(oldResourceMap, newResourceMap map[string]*schema.Resource) SchemaDiff {
```
Comment (ScottSuarez) [inline tools/diff-processor/diff/diff.go:31]: This function is quite large. Could you break this up into sub functions so it's more readable ?

**Review by ScottSuarez — CHANGES_REQUESTED** [review]

File: tools/diff-processor/rules/rules_field.go [hunk 2]
```diff
@@ -76,8 +80,12 @@ var fieldRule_BecomingRequired = FieldRule{
 }
 
 func fieldRule_BecomingRequired_func(old, new *schema.Schema, mc MessageContext) string {
+	// Ignore for removed fields
+	if new == nil {
+		return ""
+	}
 	message := mc.message
-	if !old.Required && new.Required {
+	if (old == nil || !old.Required) && new.Required {
```
[the remark this answers is not in the document]
Comment (melinath) [reply tools/diff-processor/rules/rules_field.go:88]: reverted this change in ebf0e00e9 and captured in b/300515447

File: tools/diff-processor/diff/diff.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	// If it changed to/from nil, it changed
+	oldFuncIsNil := reflect.ValueOf(oldFunc).IsNil()
+	newFuncIsNil := reflect.ValueOf(newFunc).IsNil()
+	if (oldFuncIsNil && !newFuncIsNil) || (!oldFuncIsNil && newFuncIsNil) {
+		return true
+	}
+
+	// If a func is set before and after we don't currently have a way to reliably
+	// determine whether the function changed, so we assume that it has not changed.
+	// b/300157205
```
[the remark this answers is not in the document]
Comment (melinath) [reply tools/diff-processor/diff/diff.go:229]: Possibly not.

File: tools/diff-processor/diff/diff.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+		allFields := make(map[string]struct{})
+
+		for key, _ := range flattenedOldSchema {
+			allFields[key] = struct{}{}
+		}
+
+		for key, _ := range flattenedNewSchema {
+			allFields[key] = struct{}{}
+		}
```
[the remark this answers is not in the document]
Comment (melinath) [reply tools/diff-processor/diff/diff.go:69]: I'd rather not if that's okay - it only occurs twice and it's pretty clear to read as-is.

File: tools/diff-processor/diff/diff.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+		resourceDiff.Fields = make(map[string]FieldDiff)
+		for key, _ := range allFields {
+			oldField := flattenedOldSchema[key]
+			newField := flattenedNewSchema[key]
```
[the remark this answers is not in the document]
Comment (melinath) [reply tools/diff-processor/diff/diff.go:74]: yep! https://go.dev/doc/effective_go#maps

File: tools/diff-processor/diff/diff.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+type FieldDiff struct {
+	Old *schema.Schema
+	New *schema.Schema
+}
+
+func ComputeSchemaDiff(oldResourceMap, newResourceMap map[string]*schema.Resource) SchemaDiff {
+	allResources := make(map[string]struct{})
+	schemaDiff := make(SchemaDiff)
+
```
[the remark this answers is not in the document]
Comment (melinath) [reply tools/diff-processor/diff/diff.go:34]: b/300517074
b/300516926

File: tools/diff-processor/diff/diff.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (melinath) [reply tools/diff-processor/diff/diff.go:69]: actually we'll probably need this function for other parts of this implementation - I'll add it.

[the remark this answers is not in the document]
Comment (melinath) [reply tools/diff-processor/diff/diff.go]: Resolved in 4bb8ca21d

File: tools/diff-processor/diff/diff.go (hunk 1 above)
Comment (melinath) [reply tools/diff-processor/diff/diff.go:32]: I've factored out the key union logic - I'm not sure what else I would be able to factor out, and it's only ~30 lines at this point, and it's all closely related. Let me know if there's anything specific you think would make sense to factor out.

File: tools/diff-processor/rules/rules_field.go [hunk 7]
```diff
@@ -80,12 +80,12 @@ var fieldRule_BecomingRequired = FieldRule{
 }
 
 func fieldRule_BecomingRequired_func(old, new *schema.Schema, mc MessageContext) string {
-	// Ignore for removed fields
-	if new == nil {
+	// Ignore for added / removed fields
```
Comment (ScottSuarez) [inline tools/diff-processor/rules/rules_field.go:83]: Do we want to add a ticket for this

 **adding a required field**

**Review by ScottSuarez — APPROVED** [review]

File: tools/diff-processor/rules/rules_field.go (hunk 7 above)
Comment (melinath) [reply tools/diff-processor/rules/rules_field.go:83]: added b/300515447

**Pull request merged**